src/openai/lib/azure.py

```diff
@@ -0,0 +1,349 @@
+from .._base_client import DEFAULT_MAX_RETRIES, BaseClient
+
+_deployments_endpoints = set(
+    [
+        "/completions",
+        "/chat/completions",
+        "/embeddings",
+        "/audio/transcriptions",
+        "/audio/translations",
+    ]
+)
+
+
+AzureADTokenProvider = Callable[[], str]
+AsyncAzureADTokenProvider = Callable[[], "str | Awaitable[str]"]
+_HttpxClientT = TypeVar("_HttpxClientT", bound=Union[httpx.Client, httpx.AsyncClient])
+
+
+# we need to use a sentinel API key value for Azure AD
+# as we don't want to make the `api_key` in the main client Optional
+# and Azure AD tokens may be retrieved on a per-request basis
+API_KEY_SENTINEL = "".join(["<", "missing API key", ">"])
+
+
+class MutuallyExclusiveAuthError(OpenAIError):
+    def __init__(self) -> None:
+        super().__init__(
+            "The `api_key`, `azure_ad_token` and `azure_ad_token_provider` arguments are mutually exclusive; Only one can be passed at a time"
+        )
+
+
+class BaseAzureClient(BaseClient[_HttpxClientT]):
+    def _build_request(
+        self,
+        options: FinalRequestOptions,
+    ) -> httpx.Request:
+        if options.url in _deployments_endpoints and is_mapping(options.json_data):
+            model = options.json_data.get("model")
+            if model is not None and not "/deployments" in str(self.base_url):
+                options.url = f"/deployments/{model}{options.url}"
+
+        return super()._build_request(options)
+
+
+class AzureOpenAI(BaseAzureClient[httpx.Client], OpenAI):
+    @overload
+    def __init__(
+        self,
+        *,
+        api_version: str,
+        endpoint: str,
+        deployment: str | None = None,
+        api_key: str | None = None,
+        azure_ad_token: str | None = None,
+        azure_ad_token_provider: AzureADTokenProvider | None = None,
+        organization: str | None = None,
+        timeout: float | Timeout | None | NotGiven = NOT_GIVEN,
+        max_retries: int = DEFAULT_MAX_RETRIES,
+        default_headers: Mapping[str, str] | None = None,
+        default_query: Mapping[str, object] | None = None,
+        http_client: httpx.Client | None = None,
+        _strict_response_validation: bool = False,
+    ) -> None:
+        ...
+
+    @overload
+    def __init__(
+        self,
+        *,
+        api_version: str,
+        base_url: str,
+        api_key: str | None = None,
+        azure_ad_token: str | None = None,
+        azure_ad_token_provider: AzureADTokenProvider | None = None,
+        organization: str | None = None,
+        timeout: float | Timeout | None | NotGiven = NOT_GIVEN,
+        max_retries: int = DEFAULT_MAX_RETRIES,
+        default_headers: Mapping[str, str] | None = None,
+        default_query: Mapping[str, object] | None = None,
+        http_client: httpx.Client | None = None,
+        _strict_response_validation: bool = False,
+    ) -> None:
+        ...
+
+    def __init__(
+        self,
+        *,
+        api_version: str,
```

Comment:
I think with the addition of environment variables for configuring the client, it might make sense to make `api_version` optional. Otherwise if I want to use all env vars, I'm stuck doing this and it doesn't feel great:

```python
client = openai.AzureOpenAI(api_version=None)
```

I think we could handle it similarly to how we do with e.g. `api_key` and just raise an exception if it's not provided to the client, nor by the environment variable. Thoughts?

Reply:
Agreed - having the same fallback logic as api_key seem appropriate here!

Reply:
Ah, thank you for catching this!